src/app/loadout/mod-utils.ts

```diff
@@ -408,20 +415,72 @@ function createPluggingStrategy(
 
   // For each remaining socket that won't have mods assigned,
   // return it to its default (usually "Empty Mod Socket")
+  // TO-DO: here's another thing that will fail for artificer armor.
+  // those have no singleInitialItemHash
   for (const leftoverSocket of existingModSockets) {
-    const defaultMod = defs?.InventoryItem.get(
+    const defaultMod = defs.InventoryItem.get(
       leftoverSocket.socketDefinition.singleInitialItemHash
     );
     const currentModEnergy = leftoverSocket.plugged?.plugDef.plug.energyCost?.energyCost || 0;
-    pluggingActions.push({
+    optionalRegains.push({
       socketIndex: leftoverSocket.socketIndex,
       mod: defaultMod as PluggableInventoryItemDefinition,
-      energyChange: -currentModEnergy,
+      energySpend: -currentModEnergy,
+      required: false,
     });
   }
 
-  // Sort so we free up energy before we consume it, if possible
-  return pluggingActions.sort(compareBy((res) => res.energyChange));
+  // sort lower gains first
+  optionalRegains.sort(compareBy((res) => -res.energySpend));
+
+  const operationSet: PluggingAction[] = [];
+
+  const itemTotalEnergy = item.energy!.energyCapacity;
+  let itemCurrentUsedEnergy = item.energy!.energyUsed;
+
+  // apply all required regains first
+  for (const regainOperation of requiredRegains) {
+    operationSet.push(regainOperation);
+    itemCurrentUsedEnergy += regainOperation.energySpend;
+  }
+
+  // keep looping til we have placed all desired mods
+  for (const spendOperation of requiredSpends) {
+    // while there's not enough energy for this mod,
+    while (itemCurrentUsedEnergy + spendOperation.energySpend > itemTotalEnergy) {
+      if (!optionalRegains.length) {
+        throw new Error(
+          `there's not enough energy to assign ${spendOperation.mod.displayProperties.name} to ${item.name}, but no more energy can be freed up`
+        );
+      }
```

Comment:
Surely this should never happen either, since the assignment algorithm found a combination that can fit in the armour?

Reply:
Maybe its best if we turn this off entirely in the optimiser actually? We don't care about this at all there.

Reply:
agreed but i would rather have a throw, at least for a while, than write a potentially infinite while loop

Reply:
Yeah that is fair